### DISCUSSION TITLE
How to verify JWT Token ?

### DISCUSSION THREAD

**Discussion by indrazm** [pocketbase/pocketbase#2929] [Q&A] [answered]
Hey, 
I am currently creating a middleware in NextJS for verifiying the JWT Token.
And if the Token verified, it will goes to protected route, otherwise it's going to Login page.

If i use jose library, i need the secretKey. 
How can i find the secret key for Verifiying the JWT ?

Many Thanks!

**Comment by ganigeorgiev** [comment]
To verify a PocketBase JWT auth token you can send `authRefresh()` call with the token, eg:

```js
pb.authStore.save(token, null);
await pb.collection("users").authRefresh();
```
On success the above will return the latest user data and a renewed token. On invalid or expired token it will throw an error.

> How can i find the secret key for Verifiying the JWT ?

You can't verify the PocketBase JWT auth token with just a single secret key because the token secret is constructed from 2 parts - common part that is shared among all tokens and a user unique part (for more details you can check https://github.com/pocketbase/pocketbase/issues/144#issuecomment-1186144841).

**Reply by anasaijaz** [reply]
This is kind of a faulty logic. What will happen in the case where the next.js frontend is authenticated using pb token, and the backend is supplied with the same JWT to make request on behalf of user.

If I refresh the token, wouldn't it invalidate the last JWT token and the frontend will be forced to logout. Instead of this the api should be called to get a user from pocketbase, if any user gets returned that means either the token is valid or the jwt secret is compromised

**Reply by ganigeorgiev** [reply]
@anasaijaz 

> If I refresh the token, wouldn't it invalidate the last JWT token

No. PocketBase is fully stateless and tokens are not stored on the server. Refreshing the token just generates a new JWT with updated `exp` claim.

To invalidate previously issued tokens users need to either change their password OR the admin to change the shared auth token secret.

In any case, this is a very old discussion. If you have further questions please consider opening a new one with more details about your concerns and your use case.

If you haven't already you could also check https://github.com/pocketbase/pocketbase/discussions/5313.

**Reply by anasaijaz** [reply]
Oh alright! That clears my doubt.
Thank you so much

**Comment by indrazm** [comment]
Thanks oh that's cool!

I assume that is only for client side right ? 
Is there any way to use that in server side ?

**Reply by ganigeorgiev** [reply, marked as answer]
No, the above should work also server-side. If you have an existing PocketBase JS SDK instance that you use server-side and don't want to modify its authStore state, then you can create a new one-off instance just for the `authRefresh()` call (_the object creation is very lightweight and shouldn't have impact on the performance_):

```js
const isValid = false;

try {
    // create a new one-off install from an existing one
    const temp = new PocketBase(pb.baseUrl)
    temp.authStore.save(token, null);
    // extra check to prevent unnecessary call in case the token is already expired or missing
    if (temp.authStore.isValid) {
        await temp.collection("users").authRefresh();
        isValid = true;
    }
} catch (_) {}

console.log(isValid)
```

**Reply by ganigeorgiev** [reply]
_I'll consider with JS SDK refactoring after v0.17.0 release to add a helper for the above as it seems to be commonly used especially in a SSR context._

**Comment by indrazm** [comment]
Got it. 
I finally able to save the token in NextJS middleware!
Thanks a lot.
